src/catalog/src/durable/impls/persist.rs

```diff
@@ -72,11 +74,27 @@ pub(crate) type Timestamp = mz_repr::Timestamp;
 /// `new_unchecked` is safe to call with a non-zero value.
 const MIN_EPOCH: Epoch = unsafe { Epoch::new_unchecked(1) };
 
-/// Human readable shard name.
-const SHARD_NAME: &str = "catalog";
+/// Human readable catalog shard name.
+const CATALOG_SHARD_NAME: &str = "catalog";
+/// Human readable upgrade shard name.
+const UPGRADE_SHARD_NAME: &str = "catalog_upgrade";
+
+/// Seed used to generate the persist shard ID for the catalog.
+const CATALOG_SEED: usize = 1;
+/// Seed used to generate the upgrade shard ID.
+///
+/// All state that gets written to persist is tagged with the version of the code that wrote that
+/// state. Persist has limited forward compatibility in how many version in the future a reader can
+/// read. Reading from persist updates state and the version that the state is tagged with. As a
+/// consequence, reading from persist may unintentionally fence out other readers and writers with
+/// a lower version. We use the upgrade shard to track what database version is actively deployed
+/// so readers from the future, such as the upgrade checker tool, don't accidentally fence out the
+/// database from persist. Only writable opened catalogs can increment the version in the upgrade
+/// shard.
+const UPGRADE_SEED: usize = 2;
```

Comment:
Can we include the example about running prod at `v0.X.*` and the upgrade checker at `v0.X+1.*` then `v0.X+2.*` here plz?

Reply:
Also worth including somewhere in the catalog code (maybe here) that persist guarantees that the version is non-decreasing, which sidesteps some races around fetching it and then acting on it

Reply:
Done